pkg/controller/node_availability.go

```diff
@@ -0,0 +1,125 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/klog/v2"
+	azClientSet "sigs.k8s.io/azuredisk-csi-driver/pkg/apis/client/clientset/versioned"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+type ReconcileNodeAvailability struct {
+	client                client.Client
+	namespace             string
+	controllerSharedState *SharedState
+	azVolumeClient        azClientSet.Interface
+}
+
+var _ reconcile.Reconciler = &ReconcileAzDriverNode{}
+
+func (r *ReconcileNodeAvailability) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+
+	n := &corev1.Node{}
+	err := r.client.Get(ctx, request.NamespacedName, n)
+
+	if errors.IsNotFound(err) {
+		return reconcile.Result{}, nil
+	}
+
+	if err == nil {
+		if n.Spec.Unschedulable == false {
+			//Node is schedulable, proceed to attempt creation of replica attachment
+			for r.controllerSharedState.failedReplicaRequestsQueue.queue.Len() > 0 {
+				replicaRequest := r.controllerSharedState.failedReplicaRequestsQueue.Pop()
+				r.controllerSharedState.priorityReplicaRequestsQueue.Push(*replicaRequest)
+			}
+			r.controllerSharedState.generateReplicaAttachmentRoutines(ctx, r, r.namespace)
+			return reconcile.Result{Requeue: false}, nil
+		}
+	}
+	return reconcile.Result{Requeue: false}, err
+}
+
+func NewNodeAvailabilityController(mgr manager.Manager, azVolumeClient azClientSet.Interface, namespace string, controllerSharedState *SharedState) (*ReconcileNodeAvailability, error) {
+	logger := mgr.GetLogger().WithValues("controller", "nodeavailability")
+	reconciler := ReconcileNodeAvailability{
+		namespace:             namespace,
+		client:                mgr.GetClient(),
+		controllerSharedState: controllerSharedState,
+		azVolumeClient:        azVolumeClient,
+	}
+
+	c, err := controller.New("nodeavailability-controller", mgr, controller.Options{
+		MaxConcurrentReconciles: 10,
+		Reconciler:              &reconciler,
+		Log:                     logger,
+	})
+
+	if err != nil {
+		klog.Errorf("Failed to create node availability controller. Error: %v", err)
+		return nil, err
+	}
+
+	//Predicate
+	p := predicate.Funcs{
```

Comment:
We're only interested in processing the replica priority queues on node creation and changes from unschedulable to schedulable. `CreateFunc` should unconditionally return `true` (as it does currently), `UpdateFunc` should only return true if the previous `Node.Unschedulable` state was `false` and current is now `true`, and the remaining funcs should unconditionally return `false`.